### DISCUSSION TITLE
What is `Effects.check/interrupt` in micromark?

### DISCUSSION THREAD

**Discussion by Enter-tainer** [micromark/micromark#174] [Q&A]
Hi, I'm trying to build a micromark extension to make it recognize pymdownx's admonitions and details syntax, namely

```md
!!! note

    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla et euismod
    nulla.

    Still inside the block
Out side the block
```

The behavior is

1. `???` or `!!!` starts a admonition block
2. a required space
3. until end of line, is admonition class and optional title
4. adminition block starts on next line. Each non-empty line should have at least 4 spaces. The block will end if it found the first non-empty line that is not indented.

When I'm doing this, I feel like I need to "peek" or "lookahead" future codes before I can make decision which token it belongs to. I'm not sure what `Effects.check/interrupt` can do and if they can help me out. I've checked the doc but it doesn't say much about its detailed behavior.

Also, for this example, could you give me some high level suggestion? There is a (buggy) repo mostly written by other people before and I can merely understand it and make it work. https://github.com/oi-wiki/remark-details

**Comment by wooorm** [comment]
What is “pymdownx”? Do you have a source?

> When I'm doing this, I feel like I need to "peek" or "lookahead"

I think what you want is called “attempt” in micromark jargon. You don’t need `check`/`interrupt`.
`attempt` is used a lot to parse all kinds of parts of markdown.

> Also, for this example, could you give me some high level suggestion?

My high-level suggest is to not do this.
See https://github.com/micromark/micromark#extending-markdown for more on that reasoning.
I recommend using the generic directive syntax.

Otherwise, I’d recommend, looking at the existing code you have, and also look at all our working and documented code (especially https://github.com/micromark/micromark-extension-directive)

**Reply by Enter-tainer** [reply]
> What is “pymdownx”?

I thought mkdocs-material and pymdownx is pretty popular. link here if you like https://squidfunk.github.io/mkdocs-material/reference/admonitions/

> You don’t need check/interrupt.
attempt is used a lot to parse all kinds of parts of markdown.

thank you for explaining that. is there a place i can learn what they do?

> My high-level suggest is to not do this.

I know it is anti-pattern but the reality is that i think the syntax you mentioned is not that popular. and i have to support random syntax from "other markdown libraries".

**Reply by wooorm** [reply]
> Is there a place i can learn what they do?

The code; there’s a lot of comments.

https://github.com/micromark/micromark/blob/26c38634092b164f9d43705f61265c1c7c576dad/packages/micromark-util-types/index.d.ts#L317-L330

>  but the reality is that i think the syntax you mentioned is not that popular. and i have to support random syntax from "other markdown libraries".

I wonder. A lot of code is in JavaScript. From what I know there is no JavaScript implementation of the extension you are looking for.
